enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,532 @@
+
+`CCM` metrics will be collected by the by the [prometheus-operator](https://github.com/openshift/prometheus-operator) via `ServiceMonitor`. The `Service` endpoint will be secured with generated certificate in a `Secret` by setting the `service.beta.openshift.io/serving-cert-secret-name` annotation.
+
+### Disaster Recovery
+
+In the original `KCM` implementation, where the `CCM` is currently hosted, is using a [semi-automated](https://docs.openshift.com/container-platform/4.5/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html) approach to disaster recovery, introducing the `forceRedeploymentReason` option in the managed CR, triggering the redeployment of the static pods, running the `KCM` code.
+
+`CCM` approach is simpler, as we run the `CCCMO` which manages the `Deployment`. Re-populating etcd will revert the old image, and the previous version of the `CCCMO` code will re-apply and restart the old `CCM` configuration.
+
+`CSI Migration` should be [reverted](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md#upgradedowngrade-migrateun-migrate).
+
+### Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrades from previous versions of OpenShift will look like:
+
+#### 4.6 -> 4.7
+
+Upgrade (all platforms):
+
+- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+- Users may choose to add a feature gate to enable the out-of-tree provider for OpenStack.
+
+Downgrade (any `CCM` supported platform):
+
+- No-Op:
+    - CVO cannot remove the new operator. It will sit dormant on a downgraded cluster.
+    - If a user enabled the out-of-tree provider for OpenStack via the feature gate, this cannot be undone.
+
+*OpenStack CCM will be in a tech-preview at 4.7, only a fresh cluster installation or users willing to enable the feature gate could optionally host out-of-tree CCM for OpenStack*
+
+#### Post 4.7 (in-tree to out-of-tree per provider)
+
+Each provider will be released independently as and when the providers are deemed stable enough.
+
+Prerequisites for upgrade:
+- The provider must support `CSI` for storage.
+- The provider must support `CSI Migration` from in-tree storage. `CSI Migration` must be approved (tests, CI, GA) for the provider upgrade by default.
+- `CSI` `StorageClass` should be able to become default after upgrade.
+
+Upgrade:
+
+- New `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` version updates the `kube-apiserver` and the `kube-controller-manager` and removes the `--cloud-provider` flag.
```

Comment:
how do they know for which cloud providers? The text above suggests that not all providers are ready in 4.8.